inputremapper/gui/reader.py

```diff
@@ -61,192 +63,131 @@ class Reader:
-        key_down_received = False
-
-        self._debounce_tick()
+    def read_continuously(self):
+        """poll the result pipe in regular intervals"""
+        self.read_timeout = GLib.timeout_add(30, self._read)
 
+    def _read(self):
+        """Read the messages from the helper and handle them"""
         while self._results.poll():
             message = self._results.recv()
-            event = self._get_event(message)
-            if event is None:
-                continue
 
-            gamepad = GAMEPAD in self.group.types
-            if not utils.should_map_as_btn(event, active_preset, gamepad):
+            message_type = message["type"]
+            message_body = message["message"]
+            if message_type == MSG_GROUPS:
+                self._update_groups(message_body)
                 continue
 
-            if event.value == 0:
-                logger.debug_key(event, "release")
-                self._release(event.type_and_code)
+            if message_type == MSG_EVENT:
+                if self._recording_generator:
+                    # update the generator
+                    try:
+                        self._recording_generator.send(InputEvent(*message_body))
+                    except StopIteration:
+                        self.data_bus.signal(MessageType.recording_finished)
+                        self._recording_generator = None
                 continue
+        return True
 
-            if self._unreleased.get(event.type_and_code) == event:
-                logger.debug_key(event, "duplicate key down")
-                self._debounce_start(event.event_tuple)
-                continue
-
-            # to keep track of combinations.
-            # "I have got this release event, what was this for?" A release
-            # event for a D-Pad axis might be any direction, hence this maps
-            # from release to input in order to remember it. Since all release
-            # events have value 0, the value is not used in the combination.
-            key_down_received = True
-            logger.debug_key(event, "down")
-            self._unreleased[event.type_and_code] = event
-            self._debounce_start(event.event_tuple)
-            previous_event = event
-
-        if not key_down_received:
-            # This prevents writing a subset of the combination into
-            # result after keys were released. In order to control the gui,
-            # they have to be released.
-            return None
+    def start_recorder(self) -> None:
+        """generator which yields live updated EventCombination's
 
-        self.previous_event = previous_event
-
-        if len(self._unreleased) > 0:
-            result = EventCombination(self._unreleased.values())
-            if result == self.previous_result:
-                # don't return the same stuff twice
-                return None
+        the generator will stop once all keys or inputs are released
```

Comment:
I think this docstring belongs to `_recorder`, not start_recorder`

Reply:
You are right. And the docstring needs updating the genarator actually yields nothing. It sends messages on the data_bus